### ISSUE TITLE
Type guard fails to narrow union type

### ISSUE THREAD

**Issue by altano** [Microsoft/TypeScript#1802]
This works, AS EXPECTED:
```javascript
class Cat {}
class Dog {}

function logType1(animal: Cat|Dog) {
	if (animal instanceof Cat) {
		console.log("cat");
	}
	else if (animal instanceof Dog) {
		console.log("dog");
	}
}
```

This doesn't work, which is debatable since `undefined instanceof Cat` is legal JS but I prefer the type warning since it catches errors:
```javascript
function logType2(animal: Cat|Dog|void) {
	if (animal instanceof Cat) {
		console.log("cat");
	}
	else if (animal instanceof Dog) {
		console.log("dog");
	}
}
```

This doesn't work, but I would *definitely* expect it to:
```javascript
function logType3(animal: Cat|Dog|void) {
	if (typeof animal === "undefined" || animal === null) { // edited to check for null as well, since void can be either
		console.log("no animal");
	}
	else if (animal instanceof Cat) {
		console.log("cat");
	}
	else if (animal instanceof Dog) {
		console.log("dog");
	}
}
```

See [playground](http://www.typescriptlang.org/Playground#src=class%20Cat%20{}%0Aclass%20Dog%20{}%0A%0Afunction%20logType1%28animal%3A%20Cat|Dog%29%20{%0A%09if%20%28animal%20instanceof%20Cat%29%20{%0A%09%09console.log%28%22cat%22%29%3B%0A%09}%0A%09else%20if%20%28animal%20instanceof%20Dog%29%20{%0A%09%09console.log%28%22dog%22%29%3B%0A%09}%0A}%0A%0Afunction%20logType2%28animal%3A%20Cat|Dog|void%29%20{%0A%09if%20%28animal%20instanceof%20Cat%29%20{%0A%09%09console.log%28%22cat%22%29%3B%0A%09}%0A%09else%20if%20%28animal%20instanceof%20Dog%29%20{%0A%09%09console.log%28%22dog%22%29%3B%0A%09}%0A}%0A%0Afunction%20logType3%28animal%3A%20Cat|Dog|void%29%20{%0A%09if%20%28typeof%20animal%20%3D%3D%3D%20%22undefined%22%20||%20animal%20%3D%3D%3D%20null%29%20{%0A%09%09console.log%28%22no%20animal%22%29%3B%0A%09}%0A%09else%20if%20%28animal%20instanceof%20Cat%29%20{%0A%09%09console.log%28%22cat%22%29%3B%0A%09}%0A%09else%20if%20%28animal%20instanceof%20Dog%29%20{%0A%09%09console.log%28%22dog%22%29%3B%0A%09}%0A}) for code.

As it stands, if a union type contains void, I don't know how to use `instanceof` with it.

**Comment by altano** [comment]
Worth noting that this compiles fine, which is inconsistent:

```javascript
function logType4(animal?: Cat) {
	if (animal instanceof Cat) {
		console.log("Cat");
	}
}
```
Since `void` is `null` or `undefined`, I think this behavior is inconsistent.

**Comment by danquirk** [comment]
What exactly are you trying to accomplish by putting ```void``` in the union parts?

**Comment by altano** [comment]
@danquirk in my code there's a callback function that I am not in control of, which can return a value of one of 2 types or nothing (undefined).  I'm trying to determine what type was returned by the function and I'm hitting this issue as a result.  It looks more like this:

```javascript
interface AnimalOrNothing {
    (): Dog | Cat | void;
}

var getAnimal: AnimalOrNothing = () => {
    if (Math.random() > 0.5) {
        return new Dog();
    }
    else if (Math.random() > 0.5) {
        return new Cat();
    }
    else {
        return;
    }
}

var unknown = getAnimal();
if (unknown instanceof Dog) {
    unknown.bark();
}
else if (unknown instanceof Cat) {
    unknown.pur();
}
else {

}
```
Note that I don't have control over the method that can either return a value or nothing.

**Comment by NoelAbrahams** [comment]
Since `void` is a basic type this issue looks to be related to #1785. The problem is when primitives are unioned with object types the compiler is not able to resolve `instanceof`.

**Comment by ahejlsberg** [comment]
With the fix in #1803, the `instanceof` operator now allows unions that contain primtives (including `void`), so I'm closing this bug.

**Comment by jbondc** [comment]
Isn't 'void' implicit in any type?

```javascript
interface AnimalOrNothing {
    (): string | number;
}

var getAnimal: AnimalOrNothing = () => {
    if (Math.random() > 0.5) {
        return new Dog();
    }
    else if (Math.random() > 0.5) {
        return new Cat();
    }
    else {
        return; // allowed already!
    }
}

var unknown = getAnimal();
if (unknown instanceof Dog) {
    unknown.bark();
}
else if (unknown instanceof Cat) {
    unknown.pur();
}
else {
  unknown // should narrow to void
}
// Or alternative 
if (unknown == null) {
    unknown // narrows to void
}
else if (unknown instanceof Cat) {
    unknown.pur();
}
else {
    unknown.bark();
}
```
